### PR TITLE
feat(skyrim-platform): embed cmd

### LINKED ISSUE #1516: Embed the console into the game/platform
- Makes it easier to debug logs
- If you close the console, the game closes with all background tasks
- We get the ability to copy text

image

### PR DESCRIPTION
**Pull request by VitaliyMubarakov**
closes #1516

## Embedded console
Now you can enable cmd and get all the information from the game console in a separate window.

image

### New SkyrimPlatform.ini settings

- ```Cmd = false``` - This parameter controls whether the console is enabled.

- ```CmdOffsetLeft = 0``` - This parameter controls the console window offset from the left edge of the screen.

- ```CmdOffsetTop = 720``` - This parameter controls the console window offset from the top edge of the screen.

- ```CmdWidth = 1900``` - This parameter controls the width of the console window.

- ```CmdHeight = 317``` - This parameter controls the height of the console window.
> The default values are for full hd

### Why not just use a game console?

- Makes it easier to debug logs.
- If you close the console, the game closes with all background tasks.
- We get the ability to copy text.
- If you open the game in windowed mode, you can play and watch the logs.

### CODE REVIEW

**Comment by VitaliyMubarakov** [comment]
@Pospelove check

**Comment by VitaliyMubarakov** [comment]
close https://github.com/skyrim-multiplayer/skymp/issues/1516

File: docs/release/dev/sp-embed-cmd.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+## Embedded console
```
Comment (Pospelove) [inline docs/release/dev/sp-embed-cmd.md:1]: could you please also create or update SP ini Settings docs?

File: skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.cpp [hunk 2]
```diff
@@ -23,6 +23,7 @@ struct ConsoleCommand
 };
 static std::map<std::string, ConsoleCommand> replacedConsoleCmd;
 static bool printConsolePrefixesEnabled = true;
+static WindowsConsolePrinter* windowsConsolePrinter = NULL;
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.cpp:26]: theoretically, we should use smart pointers here (like sharedptr or unique_ptr)
WindowsConsolePrinter has destructor that calls FreeConsole
destructor for static var should be called on DLL unload (automatically).

so, could you please use unique_ptr/shared_ptr here, then check what happens if you close game (via menu, via qqq).

if all good, unique_ptr/shared_ptr would be better

Please also make vars g_ prefixed. our original code violates our own style guide. (global vars should look like g_globalVar)

File: skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.cpp [hunk 3]
```diff
@@ -35,6 +36,10 @@ bool IsNameEqual(const std::string& first, const std::string& second)
 JsValue ConsoleApi::PrintConsole(const JsFunctionArguments& args)
 {
   g_printer->Print(args);
+
+  if (windowsConsolePrinter)
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.cpp:40]: please add {}

File: skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.h [hunk 4]
```diff
@@ -20,4 +20,8 @@ inline void Register(JsValue& exports)
   exports.SetProperty("setPrintConsolePrefixesEnabled",
                       JsValue::Function(SetPrintConsolePrefixesEnabled));
 }
+
+void InitCMD(const int offsetLeft, const int offsetTop, const int width,
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.h:24]: const int -> int. const bool ->bool. InitCMD -> InitCmd

File: skyrim-platform/src/platform_se/skyrim_platform/main.cpp [hunk 5]
```diff
@@ -85,6 +86,24 @@ void InitLog()
   logger::info(FMT_STRING("{} v{}"), Version::PROJECT, Version::NAME);
 }
 
+void InitCMD()
+{
+  auto settings = Settings::GetPlatformSettings();
+  const bool isCMD = settings->GetBool("Debug", "CMD", false);
+
+  if (!isCMD)
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/main.cpp:94]: isCMD => isCmd
InitCMD => InitCmd
add {} to if

**Review by Pospelove — CHANGES_REQUESTED** [review]

File: skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.cpp (hunk 3 above)
Comment (VitaliyMubarakov) [reply skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.cpp:40]: fix

File: skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.h (hunk 4 above)
Comment (VitaliyMubarakov) [reply skyrim-platform/src/platform_se/skyrim_platform/ConsoleApi.h:24]: fixed

**Comment by VitaliyMubarakov** [comment]
@Pospelove ready

**Review by Pospelove — CHANGES_REQUESTED** [review]
1. please update https://github.com/skyrim-multiplayer/skymp/blob/0714e743c9bd16f97bf2c8d9222a112cce413e72/docs/docs_skyrim_platform.md
(add link to ini_settings.md)

2. sp_ini_settings -> ini_settings

Overall seems good. Nice job! Ansgar and our whole team loves it!

**Review by Pospelove — APPROVED** [review]

**Comment by Pospelove** [comment]
awesome feature
thank u <3

**Pull request merged**